src/Coordination/KeeperStorage.cpp

```diff
@@ -1558,6 +1570,270 @@ struct KeeperStorageRemoveRequestProcessor final : public KeeperStorageRequestPr
+            {
+                auto new_responses = processWatchesImpl(it->path, watches, list_watches, Coordination::Event::DELETED);
+                responses.insert(responses.end(), std::make_move_iterator(new_responses.begin()), std::make_move_iterator(new_responses.end()));
+            }
+        }
+
+        return responses;
+    }
+
+private:
+    using SNode = typename Storage::Node;
+
+    class ToDeleteTreeCollector
+    {
+        Storage & storage;
+        int64_t zxid;
+        uint32_t limit;
+
+        uint32_t max_level = 0;
+        uint32_t nodes_observed = 1;  /// root node
+        std::unordered_map<uint32_t, std::vector<typename Storage::Delta>> by_level_deltas;
+
+        struct Step
+        {
+            String path;
+            std::variant<SNode, const SNode *> node;
+            uint32_t level;
+        };
+
+    public:
+        ToDeleteTreeCollector(Storage & storage_, int64_t zxid_, uint32_t limit_)
+            : storage(storage_)
+            , zxid(zxid_)
+            , limit(limit_)
+        {
+        }
+
+        bool collect(StringRef root_path, const SNode & root_node)
+        {
+            std::deque<Step> steps;
+
+            if (checkLimits(root_node))
+                return true;
+
+            steps.push_back(Step{root_path.toString(), &root_node, 0});
+
+            while (!steps.empty())
+            {
+                Step step = std::move(steps.front());
+                steps.pop_front();
+
+                StringRef path = step.path;
+                uint32_t level = step.level;
+                const SNode * node = nullptr;
+
+                if (auto * rdb = std::get_if<SNode>(&step.node))
+                    node = rdb;
+                else
+                    node = std::get<const SNode *>(step.node);
+
+                chassert(!path.empty());
+                chassert(node != nullptr);
+
+                if (visitRocksDBNode(steps, path, level) || visitMemNode(steps, path, level) || visitRootAndUncommitted(steps, path, *node, level))
+                    return true;
+            }
+
+            return false;
+        }
+
+        std::vector<typename Storage::Delta> extractDeltas()
+        {
+            std::vector<typename Storage::Delta> deltas;
+
+            for (ssize_t level = max_level; level >= 0; --level)
+            {
+                auto & level_deltas = by_level_deltas[static_cast<uint32_t>(level)];
+                deltas.insert(deltas.end(), std::make_move_iterator(level_deltas.begin()), std::make_move_iterator(level_deltas.end()));
+            }
+
+            return std::move(deltas);
+        }
+
+    private:
+        bool visitRocksDBNode(std::deque<Step> & steps, StringRef root_path, uint32_t level)
+        {
+            if constexpr (Storage::use_rocksdb)
+            {
+                std::filesystem::path root_fs_path(root_path.toString());
+                auto children = storage.container.getChildren(root_path.toString());
```

Comment:
Yeah, missed that scenario. I just didn't want to use getNode, because it's a guaranteed allocation on the heap if we get it not from an uncommitted state.

Reply:
You can add `exists` method to avoid it

Reply:
fixed it but in some different way